lib/sqlalchemy/sql/elements.py

```diff
@@ -3423,24 +3423,28 @@ def __init__(
             for (c, r) in new_whens
         ]
 
-        if whenlist:
-            type_ = whenlist[-1][-1].type
-        else:
-            type_ = None
-
         if value is None:
             self.value = None
         else:
             self.value = coercions.expect(roles.ExpressionElementRole, value)
 
-        self.type = cast(_T, type_)
-        self.whens = whenlist
-
         if else_ is not None:
             self.else_ = coercions.expect(roles.ExpressionElementRole, else_)
         else:
             self.else_ = None
 
+        type_ = next(
+            (
+                then.type
+                # Iterate `whens` in reverse to match previous behaviour where type of
+                # final element took priority
+                for *_, then in reversed(self.whens)
+                if not then.type._isnull
+            ),
+            self.else_.type if self.else_ is not None else None,
```

Comment:
@CaselIT CI is [failing](https://jenkins.sqlalchemy.org/job/sqlalchemy_21_gerrit/cext=cext,db=postgresql,pyv=py311/lastCompletedBuild/testReport/) with errors like:
```
TypeError: Object None associated with '.type' attribute is not a TypeEngine class or object
```

And I think it's because this line is wrong and I actually need the equivalent of:
```py
self.else_.type if self.else_ is not None else sqltype.NullType()
```

However `sqltypes` isn't available in this module and I don't think I can just add the import. What's the idiomatic way of achieving this?

Reply:
the old logic would allow type to become `NullType` in the case that none of the expressions were typed.   That is, the case "type_=None" was not actually occurring in practice.


Can we please add this patch to resolve

```diff
diff --git a/lib/sqlalchemy/sql/elements.py b/lib/sqlalchemy/sql/elements.py
index 7b93b8bf4..5dce0a295 100644
--- a/lib/sqlalchemy/sql/elements.py
+++ b/lib/sqlalchemy/sql/elements.py
@@ -3443,6 +3443,11 @@ class Case(ColumnElement[_T]):
             ),
             self.else_.type if self.else_ is not None else None,
         )
+
+        if type_ is None:
+            type_ = type_api.NULLTYPE
+
+
         self.type = cast(_T, type_)
 
     @util.ro_non_memoized_property
diff --git a/test/sql/test_case_statement.py b/test/sql/test_case_statement.py
index 6db536e69..2826b646d 100644
--- a/test/sql/test_case_statement.py
+++ b/test/sql/test_case_statement.py
@@ -13,6 +13,7 @@ from sqlalchemy import testing
 from sqlalchemy import text
 from sqlalchemy.sql import column
 from sqlalchemy.sql import table
+from sqlalchemy.sql.sqltypes import NullType
 from sqlalchemy.testing import AssertsCompiledSQL
 from sqlalchemy.testing import eq_
 from sqlalchemy.testing import fixtures
@@ -307,6 +308,23 @@ class CaseTest(fixtures.TestBase, AssertsCompiledSQL):
 
         assert isinstance(expr.type, Integer)
 
+    def test_type_of_case_expression_with_all_nulls(self):
+        expr = case(
+            (info_table.c.pk < 0, None),
+            (info_table.c.pk > 9, None),
+        )
+
+        assert isinstance(expr.type, NullType)
+
+    def test_type_of_case_expression_with_all_nulls_w_else(self):
+        expr = case(
+            (info_table.c.pk < 0, None),
+            (info_table.c.pk > 9, None),
+            else_=column('q')
+        )
+
+        assert isinstance(expr.type, NullType)
+
     def test_type_of_case_expression_with_null_case_and_else_clause(self):
         expr = case(
             (info_table.c.pk < 0, None),

```